[✨feat] InternshipAnnouncement 커스텀 예외 적용

jsoonworld opened the pull request:
# 📄 Work Description  
- `InternshipErrorCode` enum을 도입하여 마감일, 스크랩 수, 조회수에 대한 도메인 유효성 검증 메시지를 코드화함  
- `InternshipException`을 정의하여 도메인 전용 예외를 `BaseException`을 통해 처리  
- 기존 `require` 문을 제거하고, 각 도메인에서 커스텀 예외를 사용하여 검증 실패 시 명확한 예외를 던지도록 변경  

# 💭 Thoughts  
커스텀 예외를 사용하는 방식이 나중에 공통적으로 예외를 관리하거나 응답 포맷을 정형화하는 데에는 확실히 유리하다고 느꼈습니다.  
다만, 현재 도메인 규칙들이 비교적 단순한 편이어서 지금 단계에서 커스텀 예외를 도입하는 것이 과하지 않은지 고민이 되기도 했습니다.  
이 부분에 대해 유빈님의 의견을 듣고 싶습니다!

##  `require`를 사용하는 방식

**장점**  
- 간결하고 빠르게 작성 가능  
- 표준 예외(`IllegalArgumentException`)로 충분한 경우가 많음  
- 간단한 유효성 검증에는 불필요한 커스텀 클래스를 줄일 수 있음  

**단점**  
- 예외의 의미가 불명확할 수 있음 (`IllegalArgumentException`만으로는 원인을 알기 어려움)  
- 에러 코드 및 메시지의 일관성 유지가 어려움  
- API 레벨에서 정형화된 에러 응답 제공이 힘듦  

## 커스텀 예외를 사용하는 방식

**장점**  
- 예외의 의미와 출처가 명확해짐 (예: `InternshipException`)  
- 도메인 중심 설계에 적합 (비즈니스 규칙 위반을 구체적으로 표현 가능)  
- 에러 코드 및 메시지를 `enum`으로 통합 관리 가능  
- 공통 예외 핸들러에서 에러 메시지와 상태 코드를 일관되게 응답 가능  

**단점**  
- 클래스 수 증가로 인해 코드가 다소 무거워질 수 있음  
- 단순 검증에는 과할 수 있음  

➡️ 이번 변경에서는 도메인 규칙이 명확히 존재하는 영역(예: 마감일 제한, 스크랩 수 감소 제한)이므로, 명시적인 `InternshipException`을 통해 도메인 규칙을 명확하게 표현하는 방향이 더 적절하다고 판단했습니다.

# ✅ Testing Result  
![스크린샷](https://github.com/user-attachments/assets/59132108-4524-4e50-8242-5197ea5db9f8)

# 🗂 Related Issue  
- closed #48

leeeyubin committed:
feat(AuthType) : AuthType 구현

leeeyubin committed:
feat(AuthType) : AuthType 구현

leeeyubin committed:
feat(AuthErrorCode) : AuthErrorCode 구현

leeeyubin committed:
feat(AuthErrorCode) : AuthErrorCode 구현

leeeyubin committed:
feat(AuthErrorCode) : AuthErrorCode 구현

leeeyubin committed:
feat(AuthErrorCode) : AuthErrorCode 구현

leeeyubin committed:
feat(AuthErrorCode) : AuthErrorCode 구현

leeeyubin committed:
feat(AuthErrorCode) : AuthErrorCode 구현

leeeyubin committed:
feat(AuthErrorCode) : AuthErrorCode 구현

leeeyubin committed:
feat(AuthErrorCode) : AuthErrorCode 구현

leeeyubin committed:
feat(AuthErrorCode) : getPrefixMessage 함수 네이밍 변경

leeeyubin committed:
feat(AuthTypeTest) : AuthType 테스트 코드 작성

leeeyubin committed:
feat(AuthTypeTest) : ktlint 적용

leeeyubin committed:
feat(Auth) : Auth 엔티티 속성 구현

leeeyubin committed:
feat(Auth) : Auth 엔티티 속성 구현

leeeyubin committed:
feat(Auth) : Auth 엔티티 속성 구현

leeeyubin committed:
feat(Auth) : Auth 엔티티 속성 구현

leeeyubin committed:
feat(Auth) : updateRefreshToken 함수 구현

leeeyubin committed:
feat(Auth) : resetRefreshToken 함수 구현

leeeyubin committed:
feat(Auth) : var 속성 변경

leeeyubin committed:
feat(Auth) : ktlint 적용

leeeyubin committed:
feat(Banner) : Banner 엔티티 구현

jsoonworld committed:
feat(InternshipErrorCode) : 인턴십 관련 예외 코드 정의

- 마감일, 스크랩 수, 조회수에 대한 도메인 유효성 검증용 에러 코드 추가
- BaseErrorCode 인터페이스를 구현하여 상태 코드와 메시지를 일관되게 관리

jsoonworld committed:
feat(InternshipException) : 인턴십 도메인 전용 예외 클래스 구현

- InternshipErrorCode를 인자로 받아 BaseException을 상속하는 도메인 전용 예외 정의
- 인턴십 마감일, 스크랩 수, 조회수 등의 비즈니스 검증 실패 시 사용

jsoonworld committed:
refactor(InternshipDomain) : require 대신 도메인 전용 예외로 유효성 검증 방식 변경

- Deadline, ScrapCount, ViewCount에서 require를 제거하고 InternshipException으로 명확한 예외 처리 적용
- 각 예외 상황에 맞는 InternshipErrorCode를 통해 일관된 에러 메시지와 상태 코드 제공
- 도메인 규칙 위반을 명확하게 드러내는 구조로 개선

jsoonworld committed:
test(ScrapCount) : 예외 타입을 InternshipException으로 수정

- decrease() 호출 시 발생하는 예외가 IllegalArgumentException에서 InternshipException으로 변경됨에 따라 테스트 수정
- 도메인 유효성 검증 방식 변경 사항을 테스트 코드에 반영

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 추가

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 추가

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 추가

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 추가

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 추가

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 추가

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 추가

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 삭제

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 삭제

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 삭제

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 삭제

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 삭제

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 삭제

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 삭제

leeeyubin committed:
feat(AuthErrorCode) : 빈 줄 삭제

leeeyubin committed:
feat(AuthErrorCode) : EOF 수정

leeeyubin committed:
chore(AuthErrorCode) : 에러 문자 방식 수정

leeeyubin committed:
chore(AuthTypeTest) : 테스트 코드 흐름 수정

leeeyubin committed:
chore(AuthTypeTest) : ktlint 규칙 맞추기

leeeyubin committed:
chore(RefreshToken) : RefreshToken VO 구현

leeeyubin committed:
feat(AuthId) : AuthId VO 구현

leeeyubin committed:
feat(Auth) : AuthType에 EnumType 추가

leeeyubin committed:
feat(AuthId) : Ktlint 규칙에 맞게 수정

leeeyubin committed:
feat(Auth) : ktlint 규칙에 맞게 수정

leeeyubin committed:
feat(Auth) : ktlint 규칙에 맞게 수정

leeeyubin committed:
feat(Auth) : AttributeOverride 속성 추가

leeeyubin committed:
feat(ImageUrl) : ImageUrl VO 구현

leeeyubin committed:
feat(Link) : Link VO 구현

leeeyubin committed:
feat(Banner) : VO 적용

leeeyubin committed:
feat(Banner) : ktlint 적용

leeeyubin committed:
chore(ExceptionHandler) : 로그명 수정

leeeyubin committed:
chore(ExceptionHandler) : 오류 수정

leeeyubin committed:
chore(ExceptionHandler) : 불필요한 임포트 제거

leeeyubin committed:
chore(AuthErrorCode) : BaseErrorCode 상속

leeeyubin committed:
chore(AuthException) : BaseException 상속

leeeyubin committed:
chore(ExceptionHandler) : 에러 로그 수정

leeeyubin committed:
chore(ExceptionHandler) : 불필요한 임포트 제거

leeeyubin committed:
chore(ExceptionHandler) : 불필요한 임포트 제거

leeeyubin committed:
chore(ExceptionHandler) : 불필요한 임포트 제거

leeeyubin committed:
chore(ExceptionHandler) : 불필요한 임포트 제거

leeeyubin committed:
chore(AuthErrorCode) : 에러 메시지 수정

leeeyubin committed:
chore(ImageUrl) : Embeddable 어노테이션 추가

leeeyubin committed:
chore(Link) : Embeddable 어노테이션 추가

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/internshipAnnouncement/Deadline.kt:
```diff
@@ -9,8 +9,8 @@ class Deadline private constructor(
         fun from(value: LocalDate): Deadline {
-            require(value.isAfter(LocalDate.of(2025, 1, 1))) {
-                "마감일은 2025년 이후여야 합니다."
+            if (value.isAfter(LocalDate.of(2025, 1, 1)).not()) {
+                throw InternshipException(InternshipErrorCode.INVALID_DEADLINE)
```
`not` 이라는 함수는 처음 보는 것 같네요..!
`!`와 같은 의미인 걸까요?
보통 조건문에서는 `!`로 좀 더 간결하고 직관적이게 나타낼 수 있을 것 같은데 `not` 함수를 사용해주신 이유가 궁금합니다!

jsoonworld replied:
좋은 질문 감사합니다! 😊  
말씀해주신 것처럼 `not()`은 `!`와 동일한 역할을 하는 Kotlin의 확장 함수예요.  
기능적으로 차이는 없지만, 조금 더 **함수형 스타일**이나 **가독성**을 의식해서 사용해본 부분이었습니다!

다만 조건문 안에서는 말씀처럼 `!`가 훨씬 익숙하고 직관적으로 읽히는 경우가 많아서,  
이번 경우에는 `!`로 바꿔주는 게 더 명확한 표현이 될 수 있겠다는 생각이 들었어요.  
말씀해주신 내용을 참고해서 해당 부분은 `!`로 변경해보겠습니다! 🙏

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/internshipAnnouncement/ScrapCount.kt:
```diff
@@ -7,20 +7,22 @@ class ScrapCount private constructor(
     val value: Int,
 ) {
     init {
-        require(value >= MIN_VALUE) { INVALID_SCRAP_COUNT_MESSAGE }
+        if (value < MIN_VALUE) {
+            throw InternshipException(InternshipErrorCode.INVALID_SCRAP_COUNT)
+        }
     }
```
`init` 안에 들어가는 부분도 함수화를 해주면 어떤 코드인지 이해하기 편할 것 같아요!

jsoonworld replied:
말씀 듣고 보니 init 블록 안의 로직이 바로 보이지 않아서 이해하기 어려울 수도 있겠다는 생각이 들었습니다!  
안에 들어가는 로직을 메서드로 분리하면 의도도 더 잘 드러나고, 읽는 분 입장에서도 훨씬 명확해질 것 같아요.

바로 메서드로 분리해서 정리해보겠습니다 :)  
꼼꼼한 피드백 감사합니다! 🙏

leeeyubin reviewed (approved):
고민해주신 점에 대해 장단점을 잘 적어주셔서 이해하기 편했던 것 같아요!👍

저는 "`커스텀 예외`"를 적용하는 방식이 더 좋을 것 같습니다..!
현재 코드에 있어서 과해 보일지라도 추후 확장성을 고려했을 때 커스텀 예외를 통해 일관되게 응답을 해 주는 것이 더 좋을 것 같다는 의견입니다!

jsoonworld committed:
refactor(Deadline) : 함수 선언 순서를 Kotlin 컨벤션에 맞게 정렬

jsoonworld committed:
refactor(Deadline) : 마감일 최소 기준을 상수로 추출하여 의도 명확화

jsoonworld committed:
refactor(Deadline) : 마감일 기준 상수화 및 테스트 기준 정비

jsoonworld committed:
refactor(ViewCount) : Kotlin 컨벤션에 맞춰 companion object 위치 정렬

jsoonworld committed:
chore(ci) : GraalVM 액션 안정화를 위한 CI 설정 개선

jsoonworld committed:
feat(ScrapCount) : protected 생성자 추가

jsoonworld committed:
refactor(Deadline) : not() 대신 ! 사용하여 조건 가독성 개선 및 companion object 정렬

jsoonworld committed:
refactor(Deadline) : not() 대신 ! 사용하여 조건 가독성 개선 및 companion object 정렬

jsoonworld committed:
refactor(ScrapCount) : 유효성 검사 로직 메서드로 분리 및 클래스 구조 정렬

jsoonworld committed:
Merge branch 'develop' into feat/#48

jsoonworld committed:
test(DeadlineTest) : 커스텀 예외 적용에 맞춰 테스트 수정

InternshipException을 사용하는 Deadline의 변경사항에 맞춰,
예외 타입과 메시지 검증을 기존 IllegalArgumentException에서
InternshipException 및 에러 코드 메시지로 수정하였습니다.
